### ISSUE TITLE
"Latest" tag not being resolved

### ISSUE THREAD

**Issue by nikibel** [bootlin/elixir#349]
Broken link from page "U-Boot build example: TI AM335x BeagleBoneBlack wireless" ( the first one)
Course "embedded-linux-slides.pdf" Oct. 2024 
---

Request date: 2024-11-14 06:42:26.991782
Path: /u-boot//source/configs/am335x_evm_defconfig
Query string: 
Method: GET
Status code: 400 Bad Request
Error title: Error
Error details: Invalid version

**Comment by tleb** [comment]
Thanks for the report! The issue comes from Elixir not resolving the "latest" tag.

Related bug: go on Elixir, pick U-Boot from the dropdown, you get redirected to `https://elixir.bootlin.com/u-boot//source`. Again no "latest" tag is being found.

@fstachura an idea? No related patch recently, nothing special in the indexing cron logs, nothing in the Apache error logs. Looking at `Query.query('latest')`, here are some info from the prod server:

```console
$ LXR_REPO_DIR=/srv/elixir-data/u-boot/repo/ /usr/local/elixir/script.sh get-latest 0
v2024.10

$ /usr/local/elixir/venv/bin/python
Python 3.10.12 (main, Sep 11 2024, 15:47:36) [GCC 11.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import berkeleydb
>>> db = berkeleydb.db.DB()
>>> db.open("/srv/elixir-data/u-boot/data/versions.db")
>>> db.exists(b"v2024.10")
True
>>> db.exists("v2024.10")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: Bytes or Integer object expected for key,  found
```

I was wondering if it was related to the change from bsddb3 to berkeleydb, which is why I did the last test. We would have errors in the logs if that was the issue. But indeed the `tag = decode(...)` then `self.db.vers.exists(tag)` should fail(?) as decode should return a string and exists doesn't expect a string.

It seems the bug occurs on all projects but Linux, which I don't understand.

**Comment by tleb** [comment]
GitHub's interface tells it: this has been fixed in 3d81fa4e4ebc25f2ad3d6c8bc344596162cfba89.

For some unknown reason, `self.db.vers.exists('')` started returning `True`, meaning we never called `./script.sh get-latest`. Modify the code logic to always call `./script.sh get-latest` at least once (usually only once). This will also avoid one useless call to `self.db.vers.exists()`, but this is most probably negligible.
